cmd/pitr/collector/collector.go

```diff
@@ -266,63 +242,74 @@ func (c *Collector) manageBinlog(binlog pxc.Binlog) (err error) {
+	pr, pw := io.Pipe()
 
-	err = cmd.Start()
-	if err != nil {
-		return errors.Wrap(err, "run mysqlbinlog command")
-	}
+	go readBinlog(file, pw, errBuf, binlog.Name)
 
-	for {
-		time.Sleep(10 * time.Millisecond)
-		if !pr.empty {
-			break
-		}
-		stdErr, err := ioutil.ReadAll(errOut)
-		if err != nil {
-			return errors.Wrap(err, "read mysqlbinlog error output")
-		}
-		if len(stdErr) != 0 {
-			return errors.Errorf("mysqlbinlog: %s", stdErr)
-		}
-	}
-
-	err = c.storage.PutObject(binlogName, pipeBuf, -1)
+	err = c.storage.PutObject(binlogName, pr, -1)
 	if err != nil {
 		return errors.Wrapf(err, "put %s object", binlog.Name)
 	}
 
-	stdErr, err := ioutil.ReadAll(errOut)
-	if err != nil {
-		return errors.Wrap(err, "read mysqlbinlog error output")
-	}
+	log.Println("Successfully written binlog file", binlog.Name, "to s3 with name", binlogName)
 
 	err = cmd.Wait()
 	if err != nil {
 		return errors.Wrap(err, "wait mysqlbinlog command")
 	}
 
-	if stdErr != nil && string(bytes.TrimRight(stdErr, "\n")) != pxc.UsingPassErrorMessage && len(stdErr) != 0 {
-		return errors.Errorf("mysqlbinlog: %s", stdErr)
-	}
-
-	err = c.storage.PutObject(binlogName+gtidPostfix, &setBuffer, -1)
+	err = c.storage.PutObject(binlogName+gtidPostfix, &setBuffer, int64(setBuffer.Len()))
 	if err != nil {
 		return errors.Wrap(err, "put gtid-set object")
 	}
 
 	setBuffer.WriteString(set)
-	err = c.storage.PutObject(lastSetFileName, &setBuffer, -1)
+	err = c.storage.PutObject(lastSetFileName, &setBuffer, int64(setBuffer.Len()))
 	if err != nil {
 		return errors.Wrap(err, "put last-set object")
 	}
 	c.lastSet = set
 
 	return nil
 }
+
+func readBinlog(file *os.File, pipe *io.PipeWriter, errBuf *bytes.Buffer, binlogName string) {
+	b := make([]byte, 10485760) //alloc buffer for 10mb
+
+	// we need this because if binlog slow and didn't write anything
+	// to file yet, we have to skip this error and try to read again
+	// untill some data appears
+	isEmpty := true
+	for {
+		if errBuf.Len() != 0 {
+			// stop reading since we receive error from binlog command in stderr
+			pipe.CloseWithError(errors.Errorf("Error: mysqlbinlog %s", errBuf.String()))
+			return
+		}
+		n, err := file.Read(b)
+		if err == io.EOF {
+			// if we receive EOF immediatly after starting to read file
+			// we have to skip it, since data is not appeared yet
+			// if we got something already and receive EOF error - then exit
+			if isEmpty {
+				time.Sleep(10 * time.Millisecond)
+				continue
+			}
+			break
+		}
+		if err != nil && !strings.Contains(err.Error(), "file already closed") {
+			pipe.CloseWithError(errors.Wrapf(err, "Error: reading named pipe for %s", binlogName))
+			return
+		}
+		if n == 0 {
+			time.Sleep(10 * time.Millisecond)
+			continue
+		}
+		pipe.Write(b[:n])
+		isEmpty = false
+	}
+	// we need to close pipe with io.EOF because it's common way
+	// to indicate that we reached file end and it's not a error
+	pipe.CloseWithError(io.EOF)
+}
```

Comment:
can we do something like this without using buffer and `isEmpty`
```
	for {
               if errBuf.Len() != 0 {
			// stop reading since we receive error from binlog command in stderr
			pipe.CloseWithError(errors.Errorf("Error: mysqlbinlog %s", errBuf.String()))
			return
		}

                // will be read until EOF
                // error will never be EOF
		n, err := io.Copy(pipe, file)

		if err != nil {
			pipe.CloseWithError(errors.Wrapf(err, "Error: reading named pipe for %s", binlogName))
			return
		}
		
		// there was no data yet
                if n == 0 {
			time.Sleep(100 * time.Millisecond)
			continue
		}
		
		break
	}
```

Reply:
i think no because we can read nothing from file, but don't receive EOF.
Is this correct?